lib/pool/wait_queue.js

```diff
@@ -0,0 +1,136 @@
+'use strict';
+
+const errors = require('./errors');
+const PoolClosedError = errors.PoolClosedError;
+const WaitQueueFullError = errors.WaitQueueFullError;
+const WaitQueueTimeoutError = errors.WaitQueueTimeoutError;
+
+class WaitQueueMember {
+  constructor(callback) {
+    this.callback = callback;
+    this.finished = false;
+    this.timeout = null;
+  }
+
+  _finish(err, ret) {
+    if (!this.finished) {
+      this.finished = true;
+      process.nextTick(() => this.callback.call(null, err, ret));
+    }
+
+    if (this.timeout) {
+      clearTimeout(this.timeout);
+    }
+  }
+
+  success(connection) {
+    this._finish(null, connection);
+  }
+
+  failure(err) {
+    this._finish(err);
+  }
+
+  setTimeout(cb, ms) {
+    this.timeout = setTimeout(cb, ms);
+  }
+}
+
+class WaitQueue {
+  constructor(options) {
+    this._destroyed = false;
+
+    this.timeoutMS = options.waitQueueTimeoutMS || null;
+    this.maxSize = options.waitQueueSize || null;
+    this.periodMS = options.waitQueuePeriodMS || 10;
```

Comment:
Why is 10 the default?

Reply:
There isn't a real reason, I just wanted to make sure it didn't churn too much.